src/Compilers/Core/Portable/InternalUtilities/NoThrowStream.cs

```diff
@@ -0,0 +1,53 @@
+﻿
+using System;
+using System.Diagnostics;
+using System.IO;
+
+namespace Roslyn.Utilities
+{
+    internal class NoThrowStreamDisposer : IDisposable
+    {
+        private bool? _failed; // Nullable to assert that this is only checked after dispose
+        private readonly string _filePath;
+        private readonly TextWriter _writer;
+        private readonly Action<Exception, string, TextWriter> _exceptionHandler;
+
+        public Stream Stream { get; }
+
+        public bool Failed
+        {
+            get
+            {
+                Debug.Assert(_failed != null);
```

Comment:
There is an implicit contract that you call `Dispose` before calling `HasFailedToDispose`